packit/upstream.py

```diff
@@ -248,22 +248,63 @@ def create_patches(
             self.package_config.patch_generation_ignore_paths + sync_files_to_ignore
         )
 
+        linearize_history = False
+        # let's check whether some merge commits have parents past <upstream>
+        commits = self.get_commits_to_upstream(
+            upstream, add_upstream_head_commit=True, no_merge_commits=False
+        )
+        for commit in commits[1:]:  # 0 = upstream, don't check that one
+            for parent_commit in commit.parents:
+                if parent_commit not in commits:
+                    logger.info(f"Commit {commit} has a parent past {upstream}.")
+                    logger.info(
+                        "Git may be unable to construct the list of patches for complex histories."
+                    )
+                    logger.info(
+                        "We are going to linearize the history on a dedicated branch."
+                    )
```

Comment:
Would you like to read such a long line? That's why I split it to two.

Reply:
Yeah, that's a good point, though that long line in the logs would be read much less, than this block of code here.

Having a single `logger.info` call would express that three lines do belong together and would make this block more readable.

Reply:
> Yeah, that's a good point, though that long line in the logs would be read much less, than this block of code here.

How can you make such a bold statement?

I disagree here. For me UX is more important than DevExp, so I would over choose better UX than our DevExp.

On the other hand I don't follow how this is less readable than having a single logger call - once you read the sentences everything will be clear to you and since it's not complex code, the readability argument is not good enough for me.

Anyway, since this is now refactored completely, please take a look and let me know if this is still not readable enough for you and we can discuss further.

Reply:
I see. Well, I guess we might disagree regarding this.

To me readability is king, simply because good readability reduces cognitive load and gives more space and time to be spent on the important things (like good UX).

And this is not just about readability as in how the characters are laid on the screen. It's also the fact, that usually a logging event relates to one piece of information. Here there are 3 logging events for the same piece of information, basically using these 3 calls for formatting purposes.

And while I do understand your concerns that a long line might be ugly for CLI users, I'm personally more concerned about having 3 log lines for the same event in Packit Service.

And right now I'm not even sure, if the logging library is the right tool to display user facing messages in a CLI use-case. `click.echo()` seems more appropriate (?).

Anyways, if you think this is too much nitpicking, feel free to keep it as it is.

Reply:
please ignore me if I don't understand this discussion, but wouldn't the following be a compromise?

```python
logger.info(f"""Commit {commit} has a parent past {upstream} 
Git may be unable to construct the list of patches for complex histories. 
We are going to linearize the history on a dedicated branch."""
```